### ISSUE TITLE
Codist 6.6 Beta

### ISSUE THREAD

**Issue by wmjordan** [wmjordan/Codist#214]
## Download
[Codist 7690](https://github.com/wmjordan/Codist/files/9844492/Codist.zip)

[Codist 7670](https://github.com/wmjordan/Codist/files/9835198/Codist.zip)
[Codist 7668](https://github.com/wmjordan/Codist/files/9825906/Codist.zip)
[Codist 7664](https://github.com/wmjordan/Codist/files/9819440/Codist.zip)
[Codist 7655](https://github.com/wmjordan/Codist/files/9805470/Codist.zip)
[Codist 7634](https://github.com/wmjordan/Codist/files/9791434/Codist.zip)
[Codist 7622](https://github.com/wmjordan/Codist/files/9777766/Codist.zip)
[Codist 7617](https://github.com/wmjordan/Codist/files/9769945/Codist.zip)
[Codist 7608](https://github.com/wmjordan/Codist/files/9763692/Codist.zip)
[Codist 7572](https://github.com/wmjordan/Codist/files/9689622/Codist.zip)
[Codist 7556](https://github.com/wmjordan/Codist/files/9672948/Codist.zip)
[Codist 7550](https://github.com/wmjordan/Codist/files/9670055/Codist.zip)
[Codist 7525](https://github.com/wmjordan/Codist/files/9640120/Codist.zip)
[Codist 7507](https://github.com/wmjordan/Codist/files/9631973/Codist.zip)
[Codist 7456](https://github.com/wmjordan/Codist/files/9571127/Codist.zip)
[Codist 7450](https://github.com/wmjordan/Codist/files/9542240/Codist.zip)
[Codist 7434](https://github.com/wmjordan/Codist/files/9503522/Codist.zip)

## What's new
- General
  - [x] Fixed some functionalities were not working in some content types (b7456)
  - [x] Displayed method parameter names, default values and generic type argument names in C# Symbol Tool Tip (7655)
- Resource Monitor
  - [x] Added CPU, disk and memory monitors (7690)
    <img width="248" alt="image" src="https://user-images.githubusercontent.com/12182730/197446192-4dc92a48-6438-46e1-86d4-73da4f7b84ef.png">
- Super Quick Info
  - [x] Changed the alternative style for C# Quick Info to concise symbol signature (#216, #211, 7608, 7617, 7622, 7634, 7655, 7664, 7668, 7670)
    <img width="355" alt="image" src="https://user-images.githubusercontent.com/12182730/195566913-aefa59c7-6b1f-4ff4-8279-1b237e9a3146.png">
    <img width="375" alt="image" src="https://user-images.githubusercontent.com/12182730/195568352-71c9b331-3b7a-4672-af68-1737f9751f0b.png">
  - [x] Displayed type arguments for generic type (7608)
  - [x] Many enhancements and bug fixes for C# Quick Info (7655)
- Syntax Highlight
  - [x] Highlighter could starve thread pool in some cases (#215, b7450, 7550, 7556, 7572)
  - [x] Highlight customization in other textual windows were supported (#184, b7456, 7507, 7525)
  - [x] Memory leak in C# syntax highlighter (b7507, 7525, 7550)
- Smart Bar
  - [x] Images on Smart Bar were blurry if editor window was scaled (#213, b7434)
  - [x] Used proper way to determine multi-line selection (b7450)
  - [x] Added more commands to C# Smart Bar (7507)
  - [x] Changed tool tip styles (7608)
- Navigation Bar
  - [x] Changed tool tip styles (7608)

**Comment by wmjordan** [comment]
Hi @fitdev

I found that some features of Codist were not working in the newly introduced markdown editor from VS 17.4 preview 2. Beta 7456 should have this fixed. Please help download and test the beta.

During my local test, something weird occurred.
I compiled a 7452 version and installed the VSIX file. Afterward, the installation of Codist got completely broken. It even vanished from the Installed Extensions window. I recompiled it several times with VS 2017 and VS 2022 and eventually this 7456 got installed.

If the above weirdness happens to you, please let me know.

**Comment by fitdev** [comment]
Thank you for letting me know about this. I have not yet installed the latest VS Preview. Will do so next week probably. However I am not a good choice for testing markdown features since I do not use markdown in VS at all, and don't even know what features Codist has in regards to markdown. I only use Codist with C#. But I will certainly let you know if I encounter any issues with the latest beta.

**Comment by fitdev** [comment]
BTW It would be really great if you could implement a more detailed list of interfaces in the Super Quick Info tooltips, where it shows the whole interface inheritance chain for each interface. I am now working on types with like 30+ interfaces with complex hierarchies and generic parameters and it would help a lot to know exactly through which higher level interface a lower level interface like `IEnumerable<T>` for example has been inherited.

**Comment by wmjordan** [comment]
Well, understood. You situations are really complicated. 
For instance, the declared interfaces are A, B, C.
A inherits D, E, F.
B inherits D, G, H.
C inherits E, I.
F inherits H, J.
I inherits K.
J inherits K.

What do you think is the best way to show the above interfaces? Should those interfaces duplicated by inheritance be displayed multiple times?

Anyway, I will release a version 6.5.1 to address the issue #215 first and see whether the problem is resolved.

**Comment by fitdev** [comment]
Ideally I think there should be two possible views between which you can switch. Or maybe a unified view but with an option to turn a list-like view into a more detailed tree-like view.

- The simpler, list-like view is basically the way it is now but with an additional piece of information: show `(inherited through InterfaceName(s))` instead of just `(inherited)`. So, in your example, F would show that it is inherited through A; D would show that it is inherited through A and B; E would show that it is inherited through A and C; G - through B; H - through B; and so on.

- The second view should be a tree-like thing. It can be done in 2 different ways (both have their pluses and minuses):

1. A tree according to inheritance level, where the level of inheritance is represented by spacing on the left:
```
A
  D
  E
  F
    H
    J
      K
B
  D <-- Note: duplicated
  G
  H
C
  E <-- Note: duplicated
  I
    K <-- Note: duplicated
```

2. Alternatively, existing list can be made into a 2-level tree, where for each interface, there will be child node(s) that describe inheritance chains:
```
A
B
C
D
  A - D
  B - D
E
  A - E
  C - E
F
  A - F
G
  B - G
H
  B - H
I
  C - I
J
  A - F - J
K
  C - I - K
  A - F - J - K
```
(I hope I have not made a mistake, but you get the idea). I think `-` can be used as a separator, or perhaps some kind of arrow/icon to indicate inheritance direction more clearly. Also I think it is better to start the child nodes with the declared (parent-most) interface and work our way down towards the target interface, and not the other way around, so that you immediately know where the interface is coming from.

I think probably the `2.` variant is a better one, because the tree's depth is only 2 and it has less duplication.


Regarding responsiveness: At this moment I have not encountered any issues. But as I have explained, I do not multi-target, and very rarely edit files in shared projects, which is likely why I have not been affected.

**Comment by wmjordan** [comment]
Thank you for your comment. I think variant 2 is identical to the "simpler, list like view" actually.
I will check to see whether it is possible to implement that in the next version.

**Comment by fitdev** [comment]
> I think variant 2 is identical to the "simpler, list like view" actually.

I don't think so, because the current list does _not_ duplicate items, whereas my proposal does (when some interface is inherited through more than one path).

After thinking a bit more about this, I think that my variant 2 of the tree-like list would be the best. However I would add 2 more optimizations to it:

- When there is just on inheritance chain, there is no need to make the interface's node into a full tree-node, i.e. the inheritance chain can be shown in-line (as the same item) with the interface.
- When Generics are involved, it would be great if interfaces that differ only by type parameters _are grouped_. To illustrate this I have added `A : X<A>`, `B : X<B>`, `J : X<J>` to your example, with the expected presentation given below:

```
A
B
C
D
  Inherited through: A
  Inherited through: B
E
  Inherited through: A
  Inherited through: C
F (Inherited through: A)
G (Inherited through: B)
H (Inherited through: B)
I (Inherited through: C)
J (Inherited through: A - F)
K
  Inherited through: C - I
  Inherited through: A - F - J
X<T>
  X<A>: A
  X<B>: B
  X<J>: A - F - J
```

**Comment by wmjordan** [comment]
Build 7456 had introduced a terrible bug which caused [the Surrounding with Snippet command to crash VS](#217 ).
The new beta 7507 should have it fixed.

**Comment by fitdev** [comment]
Another related Super Quick Info improvement I want to propose is to reposition the `where TypeParam : ` generic constraints from the type's title to where the information about type parameters is presented as shown below:

![2022_09_27_14_49_33_](https://user-images.githubusercontent.com/8166847/192519017-900deb40-ff69-46e6-8233-37ec72ca86b9.jpg)

Because when you have many type parameters with complex constraints, it becomes quite difficult to find them in the long multi-line title.

**Comment by wmjordan** [comment]
@fitdev 

You are right.
I have the same thought about that, even before the first version of Super Quick Info was implemented, as well as #216 purposed by @laicasaane.
I'll try to re-implement the symbol title in the next version.

**Comment by wmjordan** [comment]
The previous version could still starve threads in the C# Interactive window.
Beta 7550 has rewritten the way how C# tagger postpones calling Roslyn, no more locks or blocking, asynchronous state machines will be created sparsely, which effectively reduces the number of threads concurrently created.

@fitdev and @laicasaane
Could you please take time download and test the beta and see whether it works fine on your computer. If it works well, I will work on the C# Quick Info symbol title.

**Comment by wmjordan** [comment]
Well, there's a beta 7556, which performs a little faster than 7550. However, either 7550 or 7556 can not highlight submissions in C# Interactive window as previous version did. The new versions just highlight the most recent submission and former submissions will not get highlight any more.

I found that VS inevitably leaks memory resource, after opening and closing files, by creating a new instance without extension and observe its resource usage on my machine.

The recent VS preview version also has a bug when installing extension on multiple side-by-side versions of VS. I had both VS 2017 and 2022 on my machine. If I install Codist on both instance during one installation process, the extension installer will report the installation on both instances are success, however, the result is on the contrary--Codist just gets removed on all VS.

**Comment by wmjordan** [comment]
Beta 75672 should have fixed previous problem in 7556 by caching parsed results for recent commands.

@fitdev 
Do you have ASP.NET projects on your computer?
Could you help test whether the recent betas work well on ASP.NET projects?

**Comment by fitdev** [comment]
I only have `*.razor`-flavored files, so no `*.cshtml` files if that's what you meant. But Codist did not seem to work well (poor syntax highlighting experience) the last time I checked.

**Comment by wmjordan** [comment]
*.razor files could not be improved. I installed Razor language service on my VS and found the content type of a Razor document does not appear to contain C#.

However, I haven't been working with web projects for quite a few years, and my work fellows all use VS Code to program their front-end projects. Thus I would not support C# highlighting in *.cshtml files.

**Comment by fitdev** [comment]
Just installed 7572. As you said, I had to install it 3 times for it to be installed on 2022 (I too have multiple versions, and it was only installed on the first one - 2017).

Checked razor files - no highlighting at all. It would be nice if you could try to add highlighting in at least the `@code { }` area.

**Comment by fitdev** [comment]
> Could you help test whether the recent betas work well on ASP.NET projects?

So, if you were not talking about `.cshtml` or `.razor` files, what did you mean?

**Comment by wmjordan** [comment]
I tried some Razor files with `@code{}` area. No way to hack into them at this moment.

I will release a new version in near future despite of that, then go for the Quick Info request.

**Comment by wmjordan** [comment]
@fitdev and @laicasaane

There's a new version which introduced a alternative style to display C# symbol signature in Super Quick Info (screenshot above). With that style, you can very easily spot out the name of the symbol, its member type, and its container symbol.

To activate the new style, select the "_Use alternative style_" in the _Super Quick Info/C#_ options page.

**Comment by laicasaane** [comment]
@wmjordan The **Base type** line should be broken down too
![image](https://user-images.githubusercontent.com/1594982/195323978-b5bb5df4-5afe-49d0-841a-ef43d94da033.png)

**Comment by laicasaane** [comment]
@wmjordan Some symbols are highlighted in black
![image](https://user-images.githubusercontent.com/1594982/195324307-c544ae7f-d33e-4f31-8359-377a665d292f.png)

**Comment by fitdev** [comment]
I really like the alternative QuickInfo style. Great work! However a few issues / suggestions:

1. Type parameters should include generic constraints
2. Not sure about including the namespace below the type name (because assembly and namespace are already shown at the bottom of the tooltip, so this seems to duplicate information)
3. I think instead of `.` you should just use space in cases like: `Nemapscae.class` or `Classname.field` etc.

![2022_10_12_14_56_50_](https://user-images.githubusercontent.com/8166847/195338095-26e9724e-352b-4cea-92b5-271ce429ffe1.jpg)

**Comment by laicasaane** [comment]
As stating here https://github.com/wmjordan/Codist/issues/216 I believe there should be an option to let us break the arguments of a method into several lines. This should be a separated setting.

**Comment by wmjordan** [comment]
Thank you all for the quick and reasonable feedback.

> The **Base type** line should be broken down too

This has the same solution as #220.

> Some symbols are highlighted in black

Will be fixed in the next beta.

> 1. Type parameters should include generic constraints

Thanks for reminding. Yes, I saw that and I will add it back later.

> 2\. this seems to duplicate information

You are right. Even without the namespace info at the bottom, it is easy to hover the mouse cursor onto the symbol to see its namespace. I will remove it in the next beta.


> 3\. instead of `.` you should just use space in cases like: `Nemapscae.class` or `Classname.field`

I will try that.

> let us list the arguments of a method in several lines

Yes, I am facing this problem.
![image](https://user-images.githubusercontent.com/12182730/195368827-d0482ecb-159e-4d83-adcd-326d5a586e95.png)

I am thinking whether it is possible to provide a better default behavior, good by default, instead of such an option.

**Comment by laicasaane** [comment]
> This has the same solution as https://github.com/wmjordan/Codist/issues/220.

I mean, the **Base type** section should be represented in multiple lines rather than in a single line, just as the **Interfaces** section

![image](https://user-images.githubusercontent.com/1594982/195377317-55215a4a-da3e-43d5-8c6e-264d7042e847.png)

The image will look somewhat like this:

```
SpawnPointEntity
{ } RpgWorld sealed class

Base type:
-  WorldEntity<PositionValue, DirectionValue, AvailabilityValue>
- Entity<Group, Eid, PositionValue, DirectionValue, AvailabilityValue>
- Entity<Group, Eid>
```

**Comment by wmjordan** [comment]
There's a new beta which has addressed some problems in the alternative style of C# Quick Info.

**Comment by laicasaane** [comment]
![image](https://user-images.githubusercontent.com/1594982/195543078-24bfcd55-12ae-4e6d-8e4b-f72977979415.png)

> if there are more than 5 arguments

Instead of a hard-coded number, we should have a setting for it. I suggest breaking down arguments list into separated lines not only because of the number of arguments, but also the length of an argument type name can be very long.

![image](https://user-images.githubusercontent.com/1594982/195544476-bb17e3b8-e3c2-49d0-b665-3f88e3b4ad3f.png)

**Comment by wmjordan** [comment]
Thanks for the bug report and suggestions.
On your screenshot of the Count method, it is inappropriate to break the arguments down to lines since all type names and argument names are very short. We need a better implementation for that type of scenario.

**Comment by laicasaane** [comment]
I think we can compare the total legnth of the method (the sum of lengths of the method name and every arguments) with the width of Quick Info window to decide if arguments should be broken down into lines.

And for cases like this `Count` method, we can factor a threshold for each argument length in that comparison? This threshold can be a setting too?

**Comment by wmjordan** [comment]
> And for cases like this `Count` method, we can factor a threshold for each argument length in that comparison? This threshold can be a setting too?

I am avoiding manual settings--we are facing many situations, but manual setting won't serve all well. A better algorithm should be the way to go.

I now have another strategy. It will try to best utilize the display space. But it never break the parameter type from its name. That is, the parameter type and the parameter name will always be placed horizontally side by side.
Here's the effect.
<img width="321" alt="image" src="https://user-images.githubusercontent.com/12182730/195565959-a347aa92-b578-4a8b-a3a0-c9a7d7d4ede4.png">
<img width="317" alt="image" src="https://user-images.githubusercontent.com/12182730/195566183-32bcacd8-58bb-447b-af70-58572c0fe1e7.png">

The too-long-to-fit parameter will be wrapped when necessary.
<img width="355" alt="image" src="https://user-images.githubusercontent.com/12182730/195566913-aefa59c7-6b1f-4ff4-8279-1b237e9a3146.png">

For small methods. It works as before.
<img width="221" alt="image" src="https://user-images.githubusercontent.com/12182730/195567934-2816cfb6-1f20-499c-844e-4c0e3995b024.png">

<img width="375" alt="image" src="https://user-images.githubusercontent.com/12182730/195568352-71c9b331-3b7a-4672-af68-1737f9751f0b.png">

**Comment by fitdev** [comment]
Thank you for the new beta! I really like how you now show type parameters! That's very clear! Excellent work!

A few suggestions / reminders:

1. Perhaps there is a way to combine 'returns' type and documentation?
![2022-10-13 19_07_19-](https://user-images.githubusercontent.com/8166847/195648963-02838596-0529-4da8-876a-d06d412527bb.jpg)
2. Base types still seem to be listed on a single line. Perhaps we can adapt a similar strategy with these as with method parameters? Or perhaps for ease of implementation and clarity we can always show them on different lines. Not sure. But cases like this may be complex to read:
![2022-10-13 19_11_18-](https://user-images.githubusercontent.com/8166847/195649774-2f78ba83-cc6a-4c1e-841f-6aff21c7d837.jpg)
3. The `{} Namespace type` still appears:
![2022-10-13 19_16_51-](https://user-images.githubusercontent.com/8166847/195650939-a694779e-20f3-4423-b59c-8a60f409400a.jpg)

**Comment by wmjordan** [comment]
> 1. Perhaps there is a way to combine 'returns' type and documentation?
> 5. Similar to 1, perhaps type parameters info and their documentation can be merged too:

In the current architecture of Codist, it is not so easy. I will think how to do that later.

> 2\. Base types still seem to be listed on a single line

Will be there in the next beta.

> 3\. The `{} Namespace type` still appears

Yes, I am still thinking about the way to cope with the existing symbol location setting.

> 4\. improve tooltip for type parameters themselves

Thanks for the suggestions. I will work on it.

**Comment by fitdev** [comment]
> In the current architecture of Codist, it is not so easy. I will think how to do that later.

Maybe during the tooltip creation process you can simply track that:
1) If the returns (or type parameters) appear right below the title (I think that's always the case for returns and always the case as long as there are type parameters) then set a flag that those are present.
2) When it comes to the next section - the one that normally shows documentation for returns or type parameters, simply check if appropriate flag has been set in step 1). If it _was_ set, then simply do not render the section, and instead _append_ documentation to the header.

In other words you only render the section with type params / returns documentation if the appropriate subsection is not present in the header below the symbol name. Otherwise you include documentation (if present) in the appropriate subsection below the symbol name header.

That's just my idea. Of course you know the architecture better so maybe I am wrong.

**Comment by wmjordan** [comment]
There's a problem when merging documentations into the signature part.

In some of my code, there are some members with long documentation for the `return` section (see the following screenshot).
(Yet `typeParameter` are usually short.)

<img width="380" alt="image" src="https://user-images.githubusercontent.com/12182730/195793861-0229ec95-3d4c-4547-9efe-9c4802a92457.png">


If we merge the long `return` documentation onto the signature, the documentation for the method will be pushed out of sight.

**Comment by fitdev** [comment]
Yes, I agree. It is problematic. Perhaps we can do it for type parameters only then.

Also, a suggestion for the future - it would be nice if tooltips can be made pinnable (in a similar way to how symbol lists can be pinned).

**Comment by wmjordan** [comment]
> Perhaps we can do it for type parameters only then.

Would it confuse the user why some documentations are on the signature part but others are below?
And from your screenshot, I saw that some type parameters had no documentation actually. Maybe it is possible to hide them to save some space.

> it would be nice if tooltips can be made pinnable

Yes, I am considering a rewritten C# Quick Info. It is possible to do so in the future, maybe in version 6.7.

**Comment by fitdev** [comment]
> Would it confuse the user why some documentations are on the signature part but others are below?

Personally I don't think it's a big deal. I think that if we can display a tooltip more effectively by utilizing space better it is still worth it. And yes, of course, if there is no documentation just hide it - no sense in wasting space.

> Yes, I am considering a rewritten C# Quick Info. It is possible to do so in the future, maybe in version 6.7.

That's great!

BTW: When do you think you will rework interface inheritance representation as per my suggestion above? Version 6.7 too? I think it would be great if base classes would be included there too, so that it is clear through which base classes different interfaces are inherited (in case of classes).

**Comment by wmjordan** [comment]
> rework interface inheritance representation as per my suggestion above

Your suggestion is valuable. However, I have to take some time to think about it.

**Comment by fitdev** [comment]
Another small suggestion. Perhaps you can add concrete type information for cases where members are inherited from base generic classes:

![2022-10-14 13_36_21-](https://user-images.githubusercontent.com/8166847/195827427-71eebbf7-125a-4976-96f2-7b69fd66a434.jpg)

For example here `ContainerKey` member (property) is defined in the base generic class `MutablePrefixTreeNode<...>` and the return type is specified as the generic type parameter `TPrefixElement`, which is correct for the base class, but a bit confusing for _where this member is used_ (i.e. the derived class with concrete types (or perhaps different but more constrained type parameters substituted)). In the example above the actual type is `string`. So perhaps you could provide that information somewhere like: `[ReturnsSymbol] TPrefixElement is string` or something like this. In other words in cases where type parameters come from some base class / interface, you show not only their name as defined in that base class / interface, but also show their actual concrete type (or type parameter if the class/struct/interface where they are used is generic too) based on the class/struct/interface where such members are used.

**Comment by wmjordan** [comment]
Good.
I will think about your suggestion and find a proper way to implement it.

Currently I am rolling back my VS installation to 17.4 preview 2.1, as I have just discovered that the recently preview 3 has a critical bug which causes a VS process related to Roslyn keeps consuming a lot of CPU resources.

**Comment by wmjordan** [comment]
There's a new beta which improves the display of type parameters.

After some contemplation, I decided to keep the type parameter documentation below the summary documentation.

> add concrete type information for cases where members are inherited from base generic classes/interfaces

This has been implemented in the new beta, however, except for inherited interfaces.

**Comment by fitdev** [comment]
> Currently I am rolling back my VS installation to 17.4 preview 2.1, as I have just discovered that the recently preview 3 has a critical bug which causes a VS process related to Roslyn keeps consuming a lot of CPU resources.

I am still on Preview 2 so has not noticed it. Thank you for letting me know.

**Comment by wmjordan** [comment]
There's a new beta which has many small tweaks and trivial enhancements.

**Comment by laicasaane** [comment]
There is an issue with the latest beta.

If a type is outside of any namespace (means it's in global space), Quick Info won't work.

`ClassA` in a namespace:

![2022-10-18++15-36-41++1666082201](https://user-images.githubusercontent.com/1594982/196381196-f2d232ff-2902-403b-9903-042b4d5b6575.png)

And `ClassB` in the global space:

![2022-10-18++15-38-52++1666082332](https://user-images.githubusercontent.com/1594982/196381306-9d8be4df-e286-4cc9-83f3-bcd2b0465217.png)

**Comment by wmjordan** [comment]
That's weird.
It is working on my computer.

<img width="291" alt="image" src="https://user-images.githubusercontent.com/12182730/196384720-3da2940a-1278-4763-8c92-9936b19b1c90.png">

**Comment by wmjordan** [comment]
There's a new beta version that has fixed global namespace problem. And more improvements.

**Comment by laicasaane** [comment]
Default values are not always independent constants. Many times we would want to "go to definition" by clicking on the type or constant name on Quick Info.

At the implementation site:
![2022-10-19++17-29-17++1666175357](https://user-images.githubusercontent.com/1594982/196667364-e669e99a-4c1b-4305-a650-8cce07916617.png)

At the call site:
![2022-10-19++17-29-55++1666175395](https://user-images.githubusercontent.com/1594982/196667374-aed66d59-1341-4fd9-9dfb-91d434f6ada2.png)

**Comment by wmjordan** [comment]
> Many times we would want to "go to definition" by clicking on the type or the field name on Quick Info

Yep. I want that too. Yet I need time to implement that.

Today I upgraded VS to 17.4 preview 4. Again, weird CPU consumption issue occurred. Have to downgrade again...

**Comment by wmjordan** [comment]
There's a new beta version which enables click and go on expressions on the symbol signature.

However, those symbols in the signature expression can not be syntax highlighted--too much work...

**Comment by laicasaane** [comment]
Thanks! Just a bit of info on type is lost, but as long as we can go to definition I think it's fine.

**Comment by fitdev** [comment]
Just found a small bug in QuickInfo when showing method parameters with nullable value types:

![2022_10_20_17_57_36_](https://user-images.githubusercontent.com/8166847/196984985-d2cb5408-b31c-42f1-a0d1-05be3ac3d0af.jpg)

**Comment by wmjordan** [comment]
Version 7670 has fixed the nullable type bug.

**Comment by wmjordan** [comment]
I kept the buggy 17.4 preview 4.

And today I added a function to monitor CPU as well as memory usage on the computer.
Now I can get notified when the roslyn bug occurs and CPU consumption goes high.

**Comment by fitdev** [comment]
Is the performance drop really serious on Preview 4? I am still on Preview 2. Should I wait upgrading?

**Comment by fitdev** [comment]
A small suggestion about QuickInfo Tooltips: It would be very nice if you could display number of implemented / inherited interfaces in the parentheses after `Interfaces` header: `Interfaces (2 + 12 = 14):` where the first number is the number of directly implemented/inherited interfaces, and the second number is the number of inherited interfaces (from base classes or other interfaces). Same for members: would be nice to display their count in parentheses.

**Comment by wmjordan** [comment]
> Is the performance drop really serious on Preview 4?

I think so. The roslyn bug is terrible. It is a multi-threaded related problem which can drain one CPU core on your machine just doing nonsense. I reported with stack trace of the problematic threads, yet no one in MS replied.

Having been bugged once more, I reverted again...back to preview 2.1.

> It would be very nice if you could display number of implemented / inherited interfaces in the parentheses after `Interfaces` header

It is not too big deal, but I am planning to rearrange the base type/interfaces on the Quick Info, thus this change won't live too long. I think it is possible to merge interfaces into the base types. For instance,

**Comment by fitdev** [comment]
> Having been bugged once more, I reverted again...back to preview 2.1.

Thank you for letting me know. I will avoid upgrading for now then. Hopefully they will fix it when DotNet 7 RTMs in November.

**Comment by LuohuaRain** [comment]
> I tried some Razor files with `@code{}` area. No way to hack into them at this moment.
> 
> I will release a new version in near future despite of that, then go for the Quick Info request.

Mark https://github.com/dotnet/razor/issues/8642
Maybe?
